Capture Window: Accessibility basics

reichlfl opened the pull request:
Introduces the accessibility interface to be used for controls beneath
the OrbitGlWidget, and unit tests where possible. The interface is
implemented for GlCanvas in this PR, but returns 0 valid children.

Functionality for GlCanvas, TimeGraph and Tracks will follow in a
subsequent PR.

For documentation, check OrbitQt/AccessibilityAdapter.h.

pierricgimmig commented on OrbitGl/GlCanvas.h:
```diff
@@ -7,12 +7,13 @@
 #include "Timer.h"
 
-class GlCanvas {
+class GlCanvas : public orbit_gl::GlAccessibleInterface {
```
I'm not sure about inheritance here. I'm wondering if on the OrbitGl side we shouldn't stick to a very generic  bounding box hierarchy that could be translated at the OrbitQt level to the proper Qt accessibility object hierarchy, but could also be used at the OrbitGl level for rectangle selection in the capture view, or for changing the mouse cursor for example (b/174337266, b/174833325). In this scenario, the hierarchy would be populated by a call to a "GetBoundingBoxHierarchy()" method on the GlCanvas, that would in turn query the timegraph, which would query tracks, etc., removing the need for inheritance. Maybe I'm missing something, but it seems we could solve multiple issues at once!?

reichlfl replied:
Absolutely right. I think I should get rid of this inheritance. For all child controls (TimeGraph, Track...) I actually went a different route and created new classes inheriting from the interface that receive "their" control in the constructur and just query the relevant attributes. A control then owns an instance of the class and exposes it - so similar to the seperate tree that Qt is creating, but the nodes are actually owned by the controls themselves instead of tracked in a central registry.

Don't know why I didn't do the same for GlCanvas, I'll change that.

reichlfl replied:
I changed this now, PTAL. I'm actually thinking if we should name this "Interface", since it actually has implementation, if only for the constructor and destructor. Not sure if we have a strict definition for "Interface"?

pierricgimmig reviewed (commented):
Hi @reichlfl , thanks for the change. The code looks good in general, please see minor comments. I do have a question regarding the design on the OrbitGl side however. Please see comment in GlCanvas.h.

reichlfl commented on OrbitGl/GlCanvas.h:
```diff
@@ -116,6 +117,23 @@ class GlCanvas {
+  [[nodiscard]] const orbit_gl::GlAccessibleInterface* AccessibleChild(int index) const override;
+  [[nodiscard]] const GlAccessibleInterface* AccessibleParent() const override { return nullptr; }
+
+  // These will be handled by the AccessibilityAdapter, no need to implement them.
+  [[nodiscard]] std::string AccessibleName() const override { return ""; }
+  [[nodiscard]] orbit_gl::A11yRole AccessibleRole() const override {
+    return orbit_gl::A11yRole::Grouping;
+  }
+  [[nodiscard]] orbit_gl::A11yRect AccessibleLocalRect() const override {
+    return orbit_gl::A11yRect();
+  }
+  [[nodiscard]] orbit_gl::A11yState AccessibleState() const override {
+    return orbit_gl::A11yState();
+  }
```
This is only required for GlCanvas, because it's a special case - but I also never really liked this and couldn't find a good design to avoid this. Reason is that those methods are handled by the OpenGlWidgetAccessible, but the Canvas can't know about this since it can't access it's "parent" (OpenGlWidget). It needs, however, to expose it's children (the TimeGraph)...

I could have the OpenGlWidgetAccessible handle the children directly and have it access the TimeGraph, but the OpenGlWidget returns a GlCanvas, the TimeGraph is defined in CaptureWindow...

But you're right that this isn't super nice. Let me see if I can come up with something better.

reichlfl commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,154 @@
+  orbit_gl::GlAccessibleInterfaceRegistry::Get().OnUnregistered(
+      [](orbit_gl::GlAccessibleInterface* iface) {
+        if (interface_map_.contains(iface)) {
+          auto adapter = dynamic_cast<A11yAdapter*>(interface_map_[iface]);
```
Yes - this also covers your comment below. I'm not checking if this element exists, I'm checking if this element can be cast to a A11yAdapter*. `dynamic_cast` will return a nullptr if the types don't match.

This is done to track elements that need to be destroyed when they are unregistered: A11yAdapter creates new adapters on `GetOrCreateAdapter`, and destroys them here. But there may be objects in that map that are not created by A11yAdapter (namely the OrbitGlWidgetAccessible, which is created by the Qt System) that should not be destroyed.

I could add a comment to make this clear.

pierricgimmig replied:
I'm not a huge fan of RTTI. The style guide discourages it also: https://google.github.io/styleguide/cppguide.html#Run-Time_Type_Information__RTTI_. I would favor an alternative if possible.

reichlfl replied:
Oh, I wasn't aware that our style guide is discouraging RTTI. It's quite straightforward to replace in this case, I will add a static set to `AccessibilityAdapter` that will explicitly track resources created through `GetOrCreateAdapter`, then we don't need to rely on RTTI.

reichlfl replied:
Hm. Actually I'm not sure if this is nicer, because it also contains a bit of an ugly workaround due to the fact that the destructor of `QAccessibleInterface` is protected - PTAL at the last commit and let me know if this is better or worse than the RTTI in your opinion...

reichlfl replied:
Davide's comments below lead to a better solution for this which doesn't require the ugly workaround any more, FYI!

reichlfl commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,154 @@
+      [](orbit_gl::GlAccessibleInterface* iface) {
+        if (interface_map_.contains(iface)) {
+          auto adapter = dynamic_cast<A11yAdapter*>(interface_map_[iface]);
+          if (adapter != nullptr) {
```
See above - this is not to check the element for nullptr, but to check if the dynamic_cast<> was successful.

dpallotti commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,154 @@
+OrbitGlWidgetAccessible::OrbitGlWidgetAccessible(OrbitGLWidget* widget)
+    : QAccessibleWidget(widget, QAccessible::Role::Graphic, "CaptureWindow") {
+  CHECK(widget != nullptr);
+  // TODO (freichl@) For some reason setting an accessible name for the Canvas results in a memory
```
Style is usually:
```suggestion
  // TODO(freichl): For some reason setting an accessible name for the Canvas results in a memory
```

reichlfl replied:
Disagree, style is usually to state a bug number :-) Thanks for pointing this comment out though, because I wanted to add a bug and forgot about it...

dpallotti commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+  }
+  int childCount() const override { return info_->AccessibleChildCount(); }
+  int indexOfChild(const QAccessibleInterface* child) const override;
+  [[nodiscard]] QAccessibleInterface* childAt(int x, int y) const override;
```
Why is this `[[nodiscard]]` and the others aren't?

reichlfl replied:
Good question. I removed the `[[nodiscard]]` since this method is defined by the `QAccessibleInterface` without it.

dpallotti commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,154 @@
+
+QAccessibleInterface* A11yAdapter::GetOrCreateAdapter(
+    const orbit_gl::GlAccessibleInterface* iface) {
+  if (!initialized_) {
```
I assume we don't have synchronization problems here, but this use of an effectively global variable does look worrying from that point of view.

reichlfl replied:
If we do have synchronization problems, then this whole method is problematic. At this point we shouldn't, but depending on how we use this in the future we might... but in this case, there's a lot more that needs to be looked at I think.

I wouldn't call this variable "global" since it's a private static member, but I can see the point. The reason here is that I wanted to avoid forcing an explicit call to something like `AccessibilityAdapter::Init()` which could be forgotten and would then result in leaking memory for created adapters. On the other hand, I could have added a seperate singleton class to provide the static functionality of `AccessibilityAdapter`, similar to `GlAccessibleInterfaceRegistry`, but would either make the `GetOrCreateAdapter` calls more ugly, or more costly, or both...

reichlfl replied:
Basically `std::call_once` does the same as I'm doing now, just in a concise and thread-safe manner, doesn't it? But this looks nicer, I changed it to use call_once with a locally defined static flag, this keeps it nicely hidden from the outside. No need to change the name of Init(), because it doesn't deal with the flag itself anymore.

dpallotti replied:
> Basically std::call_once does the same as I'm doing now, just in a concise and thread-safe manner, doesn't it?

Exactly, but that's why that's better.

dpallotti commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+/*Accessibility interface for OrbitGlWidget. See file documentation above for details.*/
+class OrbitGlWidgetAccessible : public QAccessibleWidget {
+ public:
+  OrbitGlWidgetAccessible(OrbitGLWidget* widget);
```
`explicit`?

reichlfl replied:
Yes, changed, thanks.

dpallotti commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+    return *reinterpret_cast<QAccessible::State*>(&state);
+  }
+
+  // vvv Statics vvv
```
Not a fan of these comments.

reichlfl replied:
I don't have a strong opinion... removed them, I think they are not needed anyways.

dpallotti commented on OrbitQt/AccessibilityTests.cpp:
```diff
@@ -0,0 +1,71 @@
+using orbit_gl::TestA11yImpl;
+
+TEST(Accessibility, CreationAndManagement) {
+  TestA11yImpl* obj = new TestA11yImpl(nullptr);
```
`unique_ptr` and `.get()`, or `optional` and take the address.

reichlfl replied:
Same as for OrbitGlAccessibilityTest.cpp - I explicitly want to test behavior on destruction and I think this makes it clearer that this is happening.

dpallotti replied:
Same as there, `.reset()` is really clear enough and I really don't want to see `new` and `delete` unless inevitable.

reichlfl replied:
Changed, as commented above

dpallotti commented on OrbitQt/AccessibilityTests.cpp:
```diff
@@ -0,0 +1,71 @@
+  // and unregister on construction and deletion. This can currently not be tested though because we
+  // cannot create a OrbitGlWidgetAccessible without an OrbitGlWidget, and this has too many
+  // dependencies to be instantiated in a unit test at this point
+  TestA11yImpl* obj = new TestA11yImpl(nullptr);
```
`unique_ptr` or `optional`.

reichlfl replied:
Agree here since I'm testing the explicit de-registration, changed.

reichlfl commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,154 @@
+}
+
+int A11yAdapter::indexOfChild(const QAccessibleInterface* child) const {
+  // This could be quite a bottleneck, I am not sure in which context
+  // and how excessive this method is actually called.
```
Thanks! But this is something we shouldn't just have in there all the time, but rather put it there to investigate performance and remove it afterwards, right?

reichlfl commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,154 @@
+QAccessibleInterface* A11yAdapter::childAt(int x, int y) const {
+  for (int i = 0; i < childCount(); ++i) {
+    auto c = child(i);
+    auto rect = c->rect();
```
Changed, thanks.

reichlfl commented on OrbitGl/OrbitGlAccessibility.h:
```diff
@@ -0,0 +1,125 @@
+  void Unregister(GlAccessibleInterface* iface);
+
+  void OnRegistered(Callback callback);
+  void OnUnregistered(Callback callback);
```
Changed to be more clear

reichlfl commented on OrbitGl/AccessibilityInterfaceMock.h:
```diff
@@ -0,0 +1,60 @@
+// found in the LICENSE file.
+
+#ifndef ORBIT_GL_TESTS_ACCESSIBILITY_INTERFACE_MOCK_H_
+#define ORBIT_GL_TESTS_ACCESSIBILITY_INTERFACE_MOCK_H_
```
Thanks for spotting this, I did some renaming and forgot to change the include guards. Fixed all of them.

reichlfl commented on OrbitGl/AccessibilityInterfaceMock.h:
```diff
@@ -0,0 +1,60 @@
+
+namespace orbit_gl {
+
+class TestA11yImpl : public orbit_gl::GlAccessibleInterface {
```
I've removed the "A11y" abbreviation everywhere now since I also was on the fence about this... Changed the name of the class to match the filename.

Regarding Fake vs Mock: I'm not sure if there's a suggested / accepted definition that we're following? If anything, isn't this a Stub? I've heard different definitions now, but "Fake" is either used as a super-set of mock & stub, or (e.g. https://martinfowler.com/articles/mocksArentStubs.html) for a class that actually behaves like the real thing. In this case, I'm exposing hard-coded information, which would make it a Stub... but I'm also verifying behavior (registration with the InterfaceRegistry)...

This sounds like I have a strong opinion, but I actually don't :-) Would be good to have a definition we use across our tests, though.

reichlfl replied:
Ok, following this definitions I agree... renamed the class and the header file.

reichlfl commented on OrbitGl/OrbitGlAccessibility.h:
```diff
@@ -0,0 +1,125 @@
+
+  orbit_qt::A11yAdapter will register itself for the "OnUnregistered" callback. See
+  OrbitQt/AccessibilityAdapter.h for more documentation.
+*/
```
I've now replaced all multi-line comments with `/* */` - usually if I use `//` for a longer comment it means I thought this was going to be a very short comment, and failed miserably in keeping it short...

reichlfl commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+GlAccessibleInterface, can find the corresponsing QAccessibleInterface (which will be implemented by
+a A11yAdapter in most cases, but we'll get there...). A11yAdapter::GetOrCreateAdapter will return an
+existing interface, or create a new adapter if needed. This usually happens as the tree is traversed
+- each A11yAdapter will query the children exposed through it's GlAccessibleInterface pointer, and
```
Ugh, I was really going strong on the `it's`... thanks for spotting this, I've changed it.

reichlfl commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+
+ See file documentation above for more details.
+ */
+class A11yAdapter : public QAccessibleInterface {
```
Yes, thanks, changed all the abbreviations.

reichlfl commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+ public:
+  A11yAdapter() = delete;
+  A11yAdapter(A11yAdapter& rhs) = delete;
+  A11yAdapter(A11yAdapter&& rhs) = delete;
```
Yes to both, thanks for spotting that. Fixed.

What do you refer to when you say it seems wrong? The deletion of the copy constructor in general, or just the wrong signature and missing deletion of the copy assignment?

reichlfl commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+
+  virtual QAccessible::State state() const override {
+    static_assert(sizeof(QAccessible::State) == sizeof(orbit_gl::A11yState));
+    auto state = info_->AccessibleState();
```
Done

reichlfl commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+  int indexOfChild(const QAccessibleInterface* child) const override;
+  [[nodiscard]] QAccessibleInterface* childAt(int x, int y) const override;
+
+  QString text(QAccessible::Text /*t*/) const override { return info_->AccessibleName().c_str(); }
```
Ah, thanks!

reichlfl commented on OrbitGl/OrbitGlAccessibility.h:
```diff
@@ -0,0 +1,125 @@
+  orbit_qt::A11yAdapter will register itself for the "OnUnregistered" callback. See
+  OrbitQt/AccessibilityAdapter.h for more documentation.
+*/
+class GlAccessibleInterfaceRegistry {
```
I'd argue they shouldn't - they are quite tightly linked as `GlAccessibleInterface` automatically registers and deregisters itself with the `GlAccessibleInterfaceRegistry` on construction and destruction (which actually may make a point that "interface" isn't the correct wording?). Also both the header and cpp file are still quite small, there's no general "one file per class" notion as it is in Java, why would you prefer to split them?

reichlfl replied:
Split this as discussed offline

reichlfl commented on OrbitGl/OrbitGlAccessibility.h:
```diff
@@ -0,0 +1,125 @@
+*/
+class GlAccessibleInterfaceRegistry {
+ public:
+  typedef std::function<void(GlAccessibleInterface*)> Callback;
```
So do I actually, not sure why I didn't here... thanks.

reichlfl commented on OrbitGl/OrbitGlAccessibility.h:
```diff
@@ -0,0 +1,125 @@
+  GlAccessibleInterface(GlAccessibleInterface& rhs) = delete;
+  GlAccessibleInterface(GlAccessibleInterface&& rhs) = delete;
+  GlAccessibleInterface& operator=(GlAccessibleInterface& rhs) = delete;
+  GlAccessibleInterface& operator=(GlAccessibleInterface&& rhs) = delete;
```
Fixed the parameters, thanks.
As stated above: Maybe I should just change the name "Interface"?

reichlfl commented on OrbitGl/OrbitGlAccessibilityTest.cpp:
```diff
@@ -0,0 +1,38 @@
+namespace orbit_gl {
+
+TEST(AccessibilityRegistry, Management) {
+  TestA11yImpl* impl = new TestA11yImpl(nullptr);
```
In this case, I chose to not use unique_ptr because I explicitely want to test the behavior on destruction (de-registration with `GlAccessibleInterfaceRegistry`), and I thought the explicit `delete` makes this more clear than just the unique_ptr going out of scope.

dpallotti replied:
Actually another and possibly better way to go would just be to use a scope. No need for dynamic memory allocation or `optional` at all.

dpallotti replied:
Although then you need to remember the old pointer so `unique_ptr` is probably just more convenient.

reichlfl replied:
I concede :D I have to agree that the `unique_ptr` and `.reset()` solution looks nicer and is as explicit as using the delete, changed it to `unique_ptr`.

reichlfl commented on OrbitGl/OrbitGlAccessibilityTest.cpp:
```diff
@@ -0,0 +1,38 @@
+  GlAccessibleInterfaceRegistry::Get().OnRegistered(registered_callback);
+  GlAccessibleInterfaceRegistry::Get().OnUnregistered(unregistered_callback);
+
+  TestA11yImpl* impl = new TestA11yImpl(nullptr);
```
And same reasoning as above.

reichlfl replied:
Changed, as commented above

reichlfl commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,154 @@
+    A11yAdapter::interface_map_;
+bool A11yAdapter::initialized_ = false;
+
+//================ A11yAdapter ==================
```
True, and I actually don't think they are needed...

reichlfl commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,154 @@
+  auto it = interface_map_.find(iface);
+  if (it != interface_map_.end()) {
+    return it->second;
+  } else {
```
Done

reichlfl commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+};
+
+/*Accessibility interface for OrbitGlWidget. See file documentation above for details.*/
+class OrbitGlWidgetAccessible : public QAccessibleWidget {
```
Again, I'd argue it shouldn't, because both classes defined in this header are needed to achieve the core functionality of this, and are documented together in this header. The same actually goes for the contents of OrbitGlAccessibility.h, but in this case I needed to make the split because one is included in OrbitGl, and this one is not.

dpallotti replied:
The fact that two classes work together doesn't mean they should be in the same file. The only reason would be if `OrbitGlWidgetAccessible` was only used internally to `AccessibilityAdapter`. If I want to use `OrbitGlWidgetAccessible` I should include `OrbitGlWidgetAccessible.h`, not like the file named after the other class.

reichlfl replied:
As discussed offline, I've moved this (and the comments with it) into the .cpp file, because there is no need to expose it in the header to begin with.

reichlfl commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+  A11yAdapter(A11yAdapter& rhs) = delete;
+  A11yAdapter(A11yAdapter&& rhs) = delete;
+
+  // vvv Overrides QAccessibleInterface vvv
```
Removed

reichlfl commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+  virtual QAccessible::State state() const override {
+    static_assert(sizeof(QAccessible::State) == sizeof(orbit_gl::A11yState));
+    auto state = info_->AccessibleState();
+    return *reinterpret_cast<QAccessible::State*>(&state);
```
Yes, thanks for pointing that out! I wasn't happy with the `reinterpret_cast` in this case because of this weird casting to pointer and back, `absl::bit_cast` is actually much nicer here.

reichlfl commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+  static QAccessibleInterface* GetOrCreateAdapter(const orbit_gl::GlAccessibleInterface* iface);
+  static void RegisterAdapter(const orbit_gl::GlAccessibleInterface* gl_control,
+                              QAccessibleInterface* qt_control) {
+    interface_map_.insert(std::make_pair(gl_control, qt_control));
```
Changed, thanks

reichlfl commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+
+ private:
+  mutable QObject dummy_;
+  A11yAdapter(const orbit_gl::GlAccessibleInterface* info) : info_(info){};
```
Indeed!

reichlfl commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,188 @@
+  mutable QObject dummy_;
+  A11yAdapter(const orbit_gl::GlAccessibleInterface* info) : info_(info){};
+
+  const orbit_gl::GlAccessibleInterface* info_ = nullptr;
```
True, I've removed this initialization.

reichlfl commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,154 @@
+
+QRect A11yAdapter::rect() const {
+  using orbit_gl::A11yRect;
+  using orbit_gl::GlAccessibleInterface;
```
No, thanks for spotting.

reichlfl commented:
Thanks for the review, @dpallotti and @pierricgimmig. I've addressed all of your comments and resolved the ones that didn't require any discussion. There's a few with which I disagree, I've commented and left them open, PTAL.

There's a lot of small fixup commits now since I needed to work through this in small chunks, sorry for that, but I want to avoid rebasing so you can see your changes since the last review.

pierricgimmig reviewed (approved):
Hi @reichlfl , thanks for the changes, please see minor comments. The main thing I'm not sure of is the use of RTTI. Also, I'm wondering if the "Accessibility" nomenclature is the right one in OrbitGl. If we end up using this system for selection/picking for example, it would be weird to go through "Accessibility" classes rather than a more generic "Bounding Box" hierarchy.  Could "Accessibility" be an "OrbitQt" only concept that would use geometry information gathered from OrbitGl? We can discuss this in person. That being said, I'm happy to submit as is and iterate if need be.

reichlfl commented on OrbitGl/GlCanvas.h:
```diff
@@ -116,6 +117,9 @@ class GlCanvas {
 
   [[nodiscard]] PickingManager& GetPickingManager() { return picking_manager_; }
 
+  // Accessibility - providing default implementations, overridden in CaptureWindow
+  [[nodiscard]] orbit_gl::GlAccessibleInterface* Accessibility() { return accessibility_.get(); }
```
No, the comment it outdated and should be removed, thanks for spotting!

reichlfl commented:
> Also, I'm wondering if the "Accessibility" nomenclature is the right one in OrbitGl. If we end up using this system for selection/picking for example, it would be weird to go through "Accessibility" classes rather than a more generic "Bounding Box" hierarchy. Could "Accessibility" be an "OrbitQt" only concept that would use geometry information gathered from OrbitGl? We can discuss this in person. That being said, I'm happy to submit as is and iterate if need be.

We can discuss that in person. But note that this is more than just bounding boxes - this also contains parts which are specific for the Microsoft UI Automation API (which Microsoft lists under the umbrella term of Accessibility: https://docs.microsoft.com/en-us/dotnet/framework/ui-automation/), like the control types and states, and it could actually be extended later to provide more accessibility features like tabstops, text for screenreaders etc, if we'll make this a priority.

I agree that a general concept of hierarchy and positioning would make sense in our GL Widgets, but then I'd go at it the other way around: Extend our classes with this general interface (sort of a super-simplified QWidget), and have the accessibility-classes use this.

dpallotti commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,200 @@
+ * QT provides support for this out of the box: Any QT Object has a corresponding
+ * QAccessibleInterface that is constructed automatically, and tracked in a central registry.
+ * QAccessibleInterfaces implement methods to access their children and parents, and this
+ *information is used to walk all objects in the accessibility tree starting from the main window.
+ *To simulate widgets underneath the capture window (such as tracks, event bars etc), the capture
+ *window widget needs to add QAccessibleInterface objects into this tree that provide information
+ *about the children, and their recursive children. The QAccessibleInterface hides all the nitty
+ *gritty details that the Windows API requires to query the accessibility tree starting from the
+ *main window handle.
```
```suggestion
 * information is used to walk all objects in the accessibility tree starting from the main window.
 * To simulate widgets underneath the capture window (such as tracks, event bars etc), the capture
 * window widget needs to add QAccessibleInterface objects into this tree that provide information
 * about the children, and their recursive children. The QAccessibleInterface hides all the nitty
 * gritty details that the Windows API requires to query the accessibility tree starting from the
 * main window handle.
```

dpallotti commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,200 @@
+ * OrbitGl defines a GlAccessibleInterface (see "OrbitGlAccessible.h") that exposes a relevant
+ * subset of accessibility information (most importantly child and parent information). Elements in
+ * OrbitGl can implement this interface. OrbitQt then defines an adapter class
+ *"AccessibilityAdapter" which implements the QAccessibleInterface, translating all calls from a
+ *GlAccessibleInterface to the Window API:
```
```suggestion
 * "AccessibilityAdapter" which implements the QAccessibleInterface, translating all calls from a
 * GlAccessibleInterface to the Window API:
```

dpallotti commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,200 @@
+ * The static methods of AccessibilityAdapter keep track of all adapters created so far and, given a
+ * GlAccessibleInterface, can find the corresponsing QAccessibleInterface (which will be implemented
+ * by a AccessibilityAdapter in most cases, but we'll get there...).
+ *AccessibilityAdapter::GetOrCreateAdapter will return an existing interface, or create a new
+ *adapter if needed. This usually happens as the tree is traversed
+ *- each AccessibilityAdapter will query the children exposed through its GlAccessibleInterface
+ *pointer, and will create new adapters for each of them as we go down the tree.
```
```suggestion
 * AccessibilityAdapter::GetOrCreateAdapter will return an existing interface, or create a new
 * adapter if needed. This usually happens as the tree is traversed
 * - each AccessibilityAdapter will query the children exposed through its GlAccessibleInterface
 * pointer, and will create new adapters for each of them as we go down the tree.
```

dpallotti commented on OrbitQt/AccessibilityAdapter.h:
```diff
@@ -0,0 +1,200 @@
+ * corresponding GlAccessibleInterface is deleted, these interfaces register themselves in the
+ * GlAccessibleInterfaceRegistry (see OrbitGlAccessibility.h), which in turn allows to register a
+ * callback on interface deletion. AccessibilityAdapter registers itself for this callback and
+ *cleans up interfaces accordingly.
```
```suggestion
 * cleans up interfaces accordingly.
```

dpallotti reviewed (commented):
I'm just going to say this: Have you at least considered putting this functionality in separate modules (I guess it would have to be one for Gl and one for Qt which is unfortunate). It seems to me that you are trying to keep everything in the same file to not add too much to the existing modules. Instead, this seems to be new functionality that doesn't really depend on the rest of the module. You could (maybe) have `Gl` depend on `GlAccessibility`, `Qt` depend on `QtAccessibility`, and `QtAccessibility` depend on `GlAccessibility`.

reichlfl commented:
> I'm just going to say this: Have you at least considered putting this functionality in separate modules (I guess it would have to be one for Gl and one for Qt which is unfortunate). It seems to me that you are trying to keep everything in the same file to not add too much to the existing modules. Instead, this seems to be new functionality that doesn't really depend on the rest of the module. You could (maybe) have `Gl` depend on `GlAccessibility`, `Qt` depend on `QtAccessibility`, and `QtAccessibility` depend on `GlAccessibility`.

As discussed yesterday, I split this into a separate module `OrbitAccessibility` (decided to leave out the "Gl" as this is not OpenGl-specific and could be used if we support a different UI framework as well).
- Split into `AccessibilityInterface`, `AccessibilityInterfaceRegistry`, and `AccessibleWidgetBridge`, each of them in their respective header files
- Tests now live in a separate module `OrbitAccessibilityTests`
- `OrbitGl` and `OrbitQt` both depend on the new module
- `AccessibilityAdapter` still lives in `OrbitQt` as it is Qt-specific

Adjusted namespaces and include guards correctly, and added all headers to the makefiles (hope I didn't miss anything).

reichlfl commented on OrbitGl/OrbitGlAccessibility.h:
```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
```
Split this as discussed offline

reichlfl commented on OrbitGl/OrbitGlAccessibility.h:
```diff
@@ -0,0 +1,157 @@
+ * Specialization of GlAccessibleInterface to bridge an OpenGl child element and its QtWidget
+ * parent, providing a default implementation for all methods.
+ */
+class GlAccessibilityBridge : public GlAccessibleInterface {
```
Split this as discussed offline

dpallotti commented on OrbitAccessibility/CMakeLists.txt:
```diff
@@ -0,0 +1,45 @@
+
+target_compile_options(OrbitAccessibility PRIVATE ${STRICT_COMPILE_FLAGS})
+
+target_sources(
+  OrbitAccessibility
+  PUBLIC AccessibleInterface.h
+         AccessibleInterfaceRegistry.h
+         AccessibleWidgetBridge.h
+         )
```
`PUBLIC` headers should be in the subdirectory `include/OrbitAccessibility` directory. Follow the example of other modules that do this.

reichlfl replied:
Thanks for pointing this out, I was following the wrong example for a new module - I've changed this and all the includes accordingly.

dpallotti commented on OrbitGl/GlCanvas.h:
```diff
@@ -5,6 +5,7 @@
 #ifndef ORBIT_GL_GL_CANVAS_H_
 #define ORBIT_GL_GL_CANVAS_H_
 
+#include "AccessibleInterface.h"
```
And then this will be included with `#include "OrbitAccessibility/AccessibleInterface.h"`.

dpallotti commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,269 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/*
```
Should some of this documentation be slightly updated to reflect the new module structure?

reichlfl replied:
They absolutely should, and I completely forgot about this, thanks. Updated the diagrams and the documentation to reflect the change in module structure and naming.

dpallotti commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,269 @@
+
+absl::flat_hash_map<const AccessibleInterface*, QAccessibleInterface*>
+    AccessibilityAdapter::interface_map_;
+absl::flat_hash_set<const QAccessibleInterface*> AccessibilityAdapter::managed_adapters_;
```
If I understand correctly, this should contain `unique_ptr`s, and the `new` and `delete` below should be replaces with the usage of `unique_ptr`. Or is there some specific problem with types? Because of the protected destructor? But why can't this contain pointers to `AccessibilityAdapter` if we know that's the type of the elements?

reichlfl replied:
The reason was that I needed to be able to perform a O(1) lookup based on the raw pointer (`AccessibilityAdapter::OnInterfaceDeleted`), so I can't store unique_ptr in this set. But I agree that we should avoid raw new and delete. This actually pointed me to a (IMO) better solution now: `managed_adapters_` is now a map from `AccessibleInterface` to `std::unique_ptr<AccessibilityAdapter>`. This map then actually owns the memory of all adapters created through `GetOrCreateAdapter`, and `interface_map_` contains the raw pointers for these and all interfaces managed elsewhere.

reichlfl commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,269 @@
+ * "AccessibilityAdapter" which implements the QAccessibleInterface, translating all calls from a
+ * AccessibleInterface to the Window API:
+ *
+ *                OrbitGl                     |                OrbitQt
```
They absolutely should, and I completely forgot about this, thanks. Updated the diagrams and the documentation to reflect the change in module structure and naming.

dpallotti commented on OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -0,0 +1,269 @@
+    return it->second;
+  }
+
+  auto adapter = std::unique_ptr<AccessibilityAdapter>(new AccessibilityAdapter(iface));
```
`std::make_unique<AccessibilityAdapter>(iface)`

reichlfl replied:
Would have been my first choice too, but this doesn't compile because the AccessibilityAdapter constructor is private.

dpallotti replied:
Ah right, this `make_unique` with private constructor tricks me every time.

dpallotti commented on OrbitAccessibility/include/OrbitAccessibility/AccessibleInterfaceRegistry.h:
```diff
@@ -0,0 +1,45 @@
+
+#include <functional>
+
+#include "AccessibleInterface.h"
```
I would have expected this to have to be `#include "OrbitAccessibility/AccessibleInterface.h", otherwise it would cause problems when this file is included from another module. But I guess I'm wrong, I'm a bit lost here.

reichlfl replied:
Hm true, I guess it works because AccessibleInterface.h is in the same folder, I've changed it for consistency.

dpallotti commented on OrbitAccessibility/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+
+target_compile_options(OrbitAccessibilityTests PRIVATE ${STRICT_COMPILE_FLAGS})
+
+target_sources(OrbitAccessibilityTests PRIVATE
+               include/OrbitAccessibility/AccessibleObjectFake.h)
```
Do you need this again? It's already in the sources of `OrbitAccessibility`.

reichlfl replied:
No, missed that, thanks

dpallotti commented on OrbitAccessibility/include/OrbitAccessibility/AccessibleObjectFake.h:
```diff
@@ -0,0 +1,59 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_GL_ACCESSIBLE_OBJECT_FAKE_H_
+#define ORBIT_GL_ACCESSIBLE_OBJECT_FAKE_H_
```
Adjust the include guards.

reichlfl replied:
Damn, I thought I double-checked all of them... apparently not. Thanks for spotting.

dpallotti reviewed (approved):
Looks good, I like it better now! Thank you for your patience.
